### PR TITLE
Deprecate the pedantic option for VOTables

### LINKED ISSUE #8887: Truly deprecate pedantic option in VO table
Follow up of #8715, which was milestoned to 4.0. Remove `pending=True` to emit `AstropyDeprecationWarning`. Also open a follow-up issue to remove `pedantic` altogether after a deprecation period.

### PR DESCRIPTION
**Pull request by eerovaher**
Closes #8887

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by pllim** [comment]
I think @bsipocz we should avoid new deprecation in LTS? So, I milestoned to 5.1, so this cannot be merged before 5.0 branching.

**Comment by bsipocz** [comment]
The addendum to APE-2 should be the guideline here, and I believe we said that deprecations are fine/preferred in LTS, and removal in a later version. The APE extension should preferably be wrapped up and merged though, so we have a document to consult.

**Comment by bsipocz** [comment]
However this is a weird case here, as `verify` is available for a while, so practically it doesn't really matter at this point when it will get its due `DeprecationWarning`.

File: astropy/io/votable/table.py [hunk 1]
```diff
@@ -61,10 +60,6 @@ def parse(source, columns=None, invalid='exception', verify=None,
         for more information. When not provided, uses the configuration setting
         ``astropy.io.votable.verify``, which defaults to 'ignore'.
 
-        .. versionchanged:: 4.0
```
Comment (bsipocz) [inline astropy/io/votable/table.py:64]: Keep this note, at least until `pedantic` fully disappears. I would even argue that the `versionadded` info, without the extra explanation is a very useful thing.

File: docs/io/votable/index.rst [hunk 2]
```diff
@@ -261,26 +260,40 @@ input.
 
 .. _pedantic-mode:
```
Comment (bsipocz) [inline docs/io/votable/index.rst:261]: maybe change the ref label, too

File: docs/io/votable/index.rst [hunk 3]
```diff
@@ -261,26 +260,40 @@ input.
...
 
-Pedantic mode
```
Comment (bsipocz) [inline docs/io/votable/index.rst:264]: Very nice find and cleanup, this got missed out in the review cycle of the original PR

**Comment by pllim** [comment]
Yes, sorry, I misremembered the policy. I moved milestone to 5.0.

Do we need to ping `pyvo` people to review this? If anyone will get hit by this warning, it would be `pyvo`, no? And maybe `astroquery` too.

**Comment by bsipocz** [comment]
yes, indeed astroquery will be hit, but it's not significant and in the worst case we can deal with it once this landed in astropy `main` (aka, the sooner the better to merge things like this before FF)

File: astropy/io/votable/table.py (hunk 1 above)
Comment (pllim) [reply astropy/io/votable/table.py:64]: Same comment as @bsipocz .

File: astropy/io/votable/tests/table_test.py [hunk 4]
```diff
@@ -286,5 +287,6 @@ def test_conf_pedantic_true(self, tmpdir):
 
             reload_config('astropy.io.votable')
 
-            with pytest.raises(VOWarning):
-                parse(get_pkg_data_filename('data/gemini.xml'))
+            with pytest.raises(AstropyDeprecationWarning):
+                with pytest.raises(VOWarning):
+                    parse(get_pkg_data_filename('data/gemini.xml'))
```
Comment (pllim) [inline astropy/io/votable/tests/table_test.py:292]: I think this works too.
```suggestion
            with pytest.raises(AstropyDeprecationWarning), pytest.raises(VOWarning):
                parse(get_pkg_data_filename('data/gemini.xml'))
```

File: docs/io/votable/index.rst [hunk 5]
```diff
@@ -261,26 +260,40 @@ input.
...
+^^^^^^^^^^^^^^^^^^
 
-Many VOTable files in the wild do not conform to the VOTable
-specification. If reading one of these files causes exceptions, you
-may turn off pedantic mode in `astropy.io.votable` by passing
-``pedantic=False`` to the `~astropy.io.votable.parse` or
-`~astropy.io.votable.parse_single_table` functions::
+Many VOTable files in the wild do not conform to the VOTable specification. You
+can set what should happen when a violation is encountered with the ``verify``
+argument, which can take three values:
```
Comment (pllim) [inline docs/io/votable/index.rst:268]:
```suggestion
keyword, which can take three values:
```

File: docs/io/votable/index.rst [hunk 6]
```diff
@@ -261,26 +260,40 @@ input.
...
+
+    * ``'ignore'`` - Attempt to parse the VOTable silently. This is the default
+      setting.
+    * ``'warn'`` - Attempt to parse the VOTable, but raise appropriate
+      :ref:`warnings`. It is possible to limit the number of warnings of the
+      same type to a maximum value using the
+      `astropy.io.votable.exceptions.conf.max_warnings
+      <astropy.io.votable.exceptions.Conf.max_warnings>` item in the
+      :ref:`astropy_config`
```
Comment (pllim) [inline docs/io/votable/index.rst:277]: To be consistent.
```suggestion
      :ref:`astropy_config`.
```

File: docs/io/votable/index.rst [hunk 7]
```diff
@@ -261,26 +260,40 @@ input.
...
+    * ``'exception'`` - Do not parse the VOTable and raise an exception.
+
+The ``verify`` argument can be used with the :func:`~astropy.io.votable.parse` or
```
Comment (pllim) [inline docs/io/votable/index.rst:280]:
```suggestion
The ``verify`` keyword can be used with the :func:`~astropy.io.votable.parse` or
```

File: docs/io/votable/index.rst [hunk 8]
```diff
@@ -261,26 +260,40 @@ input.
...
-Note, however, that it is good practice to report these errors to the
-author of the application that generated the VOTable file to bring the
-file into compliance with the specification.
+Note that ``'ignore'`` or ``'warn'``  mean that ``astropy`` will attempt to
+parse the VOTable, but if the specification has been violated then success
+cannot be guaranteed.
 
-Even with ``pedantic`` turned off, many warnings may still be omitted.
-These warnings are all of the type
-`~astropy.io.votable.exceptions.VOTableSpecWarning` and can be turned
```
Comment (pllim) [inline docs/io/votable/index.rst:282]: I see that this sentence is deleted altogether. Does this no longer apply?

File: astropy/io/votable/tests/table_test.py (hunk 4 above)
Comment (eerovaher) [reply astropy/io/votable/tests/table_test.py:292]: This would work, but it would also exceed the 80 character line length limit.

Comment (pllim) [reply astropy/io/votable/tests/table_test.py:292]: Our limit is 100 😸 

https://github.com/astropy/astropy/blob/46cfe3fc234c896e71a795c891642544ae676ded/setup.cfg#L153

File: docs/io/votable/index.rst (hunk 8 above)
Comment (eerovaher) [reply docs/io/votable/index.rst:282]: This was explaining that `pedantic=False` can emit (not omit, that's a typo) warnings, which might have been unexpected. What the `verify` options do is already explained above.

**Comment by eerovaher** [comment]
Should I add a second commit with the implemented feedback to this PR or would it be better to force-push and replace the starting commit?

**Comment by pllim** [comment]
> Should I add a second commit 

I think second commit is fine. Thanks!

File: astropy/io/votable/tests/table_test.py (hunk 4 above)
Comment (eerovaher) [reply astropy/io/votable/tests/table_test.py:292]: I pushed the second commit before seeing your answer, and I don't think the nested `with` statements here are bad enough to justify making another commit just to replace them, but I can change them if anyone requests more changes or if I'm asked to squash the commits.

**Review by pllim — APPROVED** [review]
LGTM. Thanks!

As @bsipocz , let's get this in sooner than later, so downstream packages have time to adjust before feature freeze.

**Pull request merged**